src/apps/pfu.cpp

```diff
@@ -0,0 +1,195 @@
+        return {};
+    }
+    std::vector<std::string> names;
+    struct dirent *entry;
+    struct stat st;
+    while ((entry = readdir(dir))) {
+        if (entry->d_type == DT_REG) {
+            std::string filename = player_dir + entry->d_name;
+            if (!stat(filename.c_str(), &st) && st.st_size > 0) {
+                names.emplace_back(entry->d_name);
+            } else {
+                logger.warn("Ignoring bad player file: {}", filename);
+            }
+        }
+    }
+    closedir(dir);
+    return names;
+}
+
+void upgrade_players(const std::vector<std::string> &players, Logger &logger) {
+    auto all_tasks = register_tasks();
+    auto succeeded = 0;
+    auto failed = 0;
+    const CharSaver saver;
+    for (auto &player : players) {
+        if (upgrade_player(player, all_tasks, logger, [&saver](const Char &ch) { saver.save(ch); })) {
+            succeeded++;
+        } else {
+            failed++;
+        }
+    }
+    logger.info("Processed characters: {} succeeded, {} failed.", succeeded, failed);
+}
+
+std::optional<Time> try_parse_login_at(std::string &login_at) {
+    if (login_at.empty()) {
+        return std::nullopt;
+    }
+    std::tm tm_login;
+    // Includes older formats used for login time.
+    for (const auto format : {"%Y-%m-%d %H:%M:%SZ", "%Y-%m-%d %H:%M:%S", "%a %b %d %H:%M:%S %Y"}) {
+        memset(&tm_login, 0, sizeof(std::tm));
+        if (strptime(login_at.c_str(), format, &tm_login)) {
+            return Clock::from_time_t(mktime(&tm_login));
+        }
+    }
+    return std::nullopt;
+}
+
+// Reset/repair Char attributes.
+ResetModifiableAttrs::ResetModifiableAttrs(CharVersion version) : UpgradeTask(version) {}
+
+// This method replaces the old handler.cpp:reset_char(). Originally it was executed
+// every time a char logged in and it reset the char's armour and attributes
+// to a healthy starting value before applying modifiers from spell and object "affects". Unclear why
+// this was introduced, but it was probably to fix "attribute drift" e.g. where the modifiers of a spell affect
+// got changed in the code without retrospectively adjusting the player's stats.
+void ResetModifiableAttrs::execute(Char &ch) const {
+    OBJ_DATA *obj;
+    // Restore the character to his/her true condition
+    // All of the attributes reset here are ones that can be modified in AFFECT_DATA.cpp
+    ranges::fill(ch.mod_stat, 0);
+    ch.max_hit = ch.pcdata->perm_hit;
+    ch.max_mana = ch.pcdata->perm_mana;
+    ch.max_move = ch.pcdata->perm_move;
+    // #216 -1 is initial value for armour now
+    ranges::fill(ch.armor, -1);
+    ch.hitroll = 0;
+    ch.damroll = 0;
+    ch.saving_throw = 0;
+    ch.sex = ch.pcdata->true_sex;
+    // add back object effects
+    for (auto loc = 0; loc < MAX_WEAR; loc++) {
+        obj = get_eq_char(&ch, loc);
+        if (!obj)
+            continue;
+        for (size_t i = 0; i < ch.armor.size(); i++)
```

Comment:
(not actionable) Darn ... C++ needs a python-like `enumerate()` which lets you get the index too. Elsewhere we use `for (auto &x : armor)` but this needs the index! annoying :)